DRIVERS-2672 Add OIDC machine workflow spec. (#1471)
mongodb/specifications: source/auth/auth.rst

.. note::
    Five minutes was chosen based on the AWS documentation for `IAM roles for EC2 <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html>`_ : "We make new credentials available at least five minutes before the expiration of the old credentials". The intent is to have some buffer between when the driver fetches the credentials and when the server verifies them.

MONGODB-OIDC
~~~~~~~~~~~~

:since: 7.0 Enterprise

MONGODB-OIDC authenticates using an `OpenID Connect (OIDC)
<https://openid.net/specs/openid-connect-core-1_0.html>`_ access token. Most
OIDC Identity Providers (*IdP*) support two identity federation flows:

- Workload Identity Federation - Also called the "machine authentication flow",
  this flow is intended to authorize access for software and services without
  requiring human interaction.

  TODO: Is there a 1:1 corresponding OIDC spec flow?

- Workforce Identity Federation - Also called the "human authentication flow",
  this flow is intended to authorize access for end users and typically requires
  human interaction to complete (e.g. clicking a button in a browser).

>   TODO: Is there a 1:1 corresponding OIDC spec flow?

Reviewer: No, all authentication flows described in the OIDC spec are human authentication flows. I’d argue that it’s a bit confusing to refer to workload/workforce identity federation as “machine authentication flow”/“human authentication flow”, neither of these are really single auth flows in the sense of OIDC.

Author: Agreed, I was trying to relate the driver process to public OIDC documentation as much as possible, but I realized that was difficult because we're really satisfying a MongoDB auth mechanism that is based on OIDC, but is not an OIDC flow itself. I've updated the document to remove the explicit separation of "machine" and "human" flows, instead assuming the "machine" flow is the normal use case for "MONGODB-OIDC", and listing the "human" flow as a separate section that most drivers can ignore.

>   TODO: Is there a 1:1 corresponding OIDC spec flow?

Reviewer: No, all authentication flows described in the OIDC spec are human authentication flows. I’d argue that it’s a bit confusing to refer to workload/workforce identity federation as “machine authentication flow”/“human authentication flow”, neither of these are really single auth flows in the sense of OIDC.

<https://openid.net/specs/openid-connect-core-1_0.html>`_ access token.

TODO: Add background on what part of the auth flow drivers play specifically.

`MongoCredential`_ Properties
`````````````````````````````

username
    MUST NOT be specified.

source
    MUST be "$external". Defaults to ``$external``.

password
    MUST NOT be specified.

mechanism
    MUST be "MONGODB-OIDC"

mechanism_properties
    PROVIDER_NAME
        Drivers MUST allow the user to specify the name of the OIDC service to
        use to obtain credentials. MUST be one of ["aws"]. If PROVIDER_NAME is
        given and a Custom provider callback is configured, the driver MUST
        raise an error.

Supported Service Providers
```````````````````````````

AWS
___

TODO: Add.

Custom
______

Drivers MUST allow the user to integrate with other OIDC providers by
implementing a custom callback that returns an OIDC token. Drivers MUST provide
a way for the callback to be either automatically canceled, or to cancel itself.
This can be as a timeout argument to the callback, a cancellation context passed
to the callback, or some other language-appropriate mechanism.  The timeout

> deadline MUST be the same as the connection establishment deadline.

Author: Updated to describe timeout concepts and values consistent with the CSOT spec.

Callbacks can be synchronous and/or asynchronous, depending on the driver
and/or language.  Asynchronous callbacks should be preferred when other
operations in the driver use asynchronous functions.

The driver MUST pass the following information to the callback:

> - ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.

Author: Renamed.

- ``mechanismProperties``: All key-value pairs from the
  AUTH_MECHANISM_PROPERTIES parameter.
- ``invalidatedToken``: Optional. The previous token that was invalidated after

>   receiving a ``ReauthenticationRequired`` error.

Author: We've decided to manage the token cache in the driver, so passing the invalidated token to the callback is no longer required.

> The callback MUST return an OIDC access token in JWT format.

Reviewer 2: Is this intended to be a documentation requirement? (Drivers could enforce this by introducing a type that validates user input, but I believe we were fine with returning strings.)

Reviewer: Generally speaking, it’s strongly recommended for clients to not inspect access tokens themselves, only for resource servers. For example, at some theoretical point in the future, MongoDB servers could add support for opaque access tokens, which would be incompatible with added JWT validation in the driver.

Author: This was intended to be a documentation requirement to inform users how to implement the callback. I will update it to clarify that.

Author: Rewrote this section to only describe what the callback API must support.

The signature of the callback is up to the driver's discretion, but the driver
MUST ensure that additional optional parameters can be added to the callback
signature in the future. An example might look like:

.. code:: typescript

  interface CallbackParameters {
      // All key-value pairs from the AUTH_MECHANISM_PROPERTIES parameter.
      mechanismProperties: Object;

      // The refresh token, if applicable, to be used by the callback to request a new token from the issuer.

>       invalidatedToken: Optional<string>;

Reviewer 2: I think this comment is incorrect. The invalidated token is the "previous" accessToken (that is, the accessToken associated with the connection that has received a 391-reauth).

The user, who is expected to handle caching and locking on their end, would use this to see if the cache is invalidated (that is, if the cache's token is the same as this invalid token, the cached refreshToken should then be used to communicate with the issuer).

We might consider renaming this to `invalidatedAccessToken`, since this might be confusing for users as well.

Author: I believe there's no longer a reason to pass the invalidated token into the callback because it was only needed to allow the callback author to manage their cache. Since we're handling caching in the driver, it shouldn't be necessary, so I'm planning to remove it.

Author: Removed.

  }

.. code:: typescript

>   function oidcToken(timeoutMS: int, params CallbackParameters): string

Author: 1. That makes sense, I'll move the timeout parameter to the parameters object/struct.
2. That's a good idea and is already relevant because we want callbacks to be able to return a token expiry time, if available, to help caching work more correctly. I'll change the return type to an object/struct.

>   function oidcToken(timeoutMS: int, params CallbackParameters): string

Author: Moved the timeout param to the `OIDCCallbackParams` and renamed it to `callbackTimeoutMS`. Also made the return type an object/struct instead of a single value.

Conversation
````````````

As an example, given the OIDC token JWT string "abcd1234", the SASL conversation looks like:

| C: :javascript:`{saslStart: 1, mechanism: "MONGODB-OIDC", payload: BinData(0, BSON({"jwt": "abcd1234"}))}`
| S: :javascript:`{conversationId : 1, payload: BinData(0, "..."), done: true, ok: 1}`

Where the sent ``payload`` is a generic binary blob containing a BSON-encoded document ``{"jwt": "abcd1234"}``.

Access Token Caching
````````````````````

Drivers MUST cache the Access Token used to authenticate a connection on the
connection object. Additionally, drivers MUST cache the most recent access Token
on the ``MongoClient``. If any operation fails with ``ReauthenticationRequired``
(391), the driver MUST clear the access Token cached on the connection object
and and reauthenticate the connection as described in ``Reauthentication``
section below.

Speculative Authentication
``````````````````````````

Drivers MUST implement speculative authentication for MONGODB-OIDC during the
``hello`` handshake. If there is a cached Access Token on the ``MongoClient``,

> use it for authentication. Otherwise, call the configured workload callback to

Reviewer 3: I thought we were not calling any callbacks during speculative auth, since it could still be a lengthy process.

Author: My understanding is that the _intended_ use case of the custom machine-to-machine callback is to allow users to integrate with OIDC providers that the driver doesn't have built-in support for (using a machine-to-machine auth flow). In that case, it _should_ behave similarly to a built-in provider (e.g. AWS or Azure) and should be significantly faster than the "human" callback. However, we don't control the implementation, so a custom callback could take several seconds (it's subject to CSOT, so technically could take up to 10-30s), which is longer than we'd expect to wait for most other auth mechanisms.

Should the "no callbacks during speculative auth" rule be extended to the custom machine-to-machine callback? Is there a case where a slow callback is a problem for creating the speculative auth payload (i.e. the initial `hello` command) but not a problem for post-`hello` auth commands (e.g. `saslStart`)?

Reviewer 2: I think the original problem was that we were worried about delays for non-authenticated hellos (ping), but we decided that this was not an issue. Later, we agreed that a server response might be missing a `speculativeAuthenticate`, and that this might be either because spec auth failed, or because it was removed by (for example) a proxy. From some of my Java notes:

> When the process is initiated during the initial handshake, speculative auth will be used. It is possible to know that speculative auth succeeded, but no reliable way to know it failed. Therefore, callbacks should not be invoked during the initial handshake.

At the time, it must have been more obvious to me why being unable to determine failure implied that (apparently all) callbacks were unworkable during initial auth (though using a cached token was ok - though, now I think we cannot tell the difference?), but now I do not immediately see why. But perhaps someone else can.

Reviewer 2: If the initial auth succeeded, but `speculativeAuthenticate` was absent, then we would redundantly invoke the callback a second time (since the absence implies that it might have failed). We should not invoke callbacks spuriously, so callbacks should not be invoked during speculative auth.

Author: I believe the redundant callback invocation problem is resolved by the updated cache management strategy for the human auth flow. The new logic assumes a missing specauth response means nothing and continues with non-speculative authentication with the cached access token. If the non-speculative authentication fails with the cached access token, then call for a new access token, first using a refresh token, then getting new `IdpInfo` and calling the human callback.

I'll go back through the proposed algorithms and make sure they're consistent, but I think we can remove the requirement to skip the callback during speculative authentication.

Author: I simplified the speculative auth behavior to only apply to cached access tokens. The result is that no new access tokens are fetched by any means (callback, built-in, etc) during speculative auth.

----------------------------

`MongoCredential`_ Properties
`````````````````````````````
username
    MUST NOT be specified.

source
    MUST be "$external". Defaults to ``$external``.

password
    MUST NOT be specified.

mechanism
    MUST be "MONGODB-OIDC"

mechanism_properties
    PROVIDER_NAME
        Drivers MUST allow the user to specify the name of the OIDC service to
        use to obtain credentials. MUST be one of ["aws"].

Interfaces
``````````

Custom Callback
```````````````
Drivers MUST provide a way for the callback to be either
automatically canceled, or to cancel itself.  This can be as a timeout argument
to the callback, a cancellation context passed to the callback, or some other
language-appropriate mechanism.  The timeout deadline MUST be the same as the
connection establishment deadline (TODO: Does that make sense?).

Callbacks can be synchronous and/or asynchronous, depending on the driver
and/or language.  Asynchronous callbacks should be preferred when other
operations in the driver use asynchronous functions.

The driver MUST pass the following information to the callback:

- ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.

> - ``mechanismProperties``: All key-value pairs from the

Reviewer 4: This seems odd.  The driver will pass `["PROVIDER_NAME": "aws"]` but will have no way of enforcing that "aws" is actually used by the callback. Also, isn't the callback itself specified as an auth mechanism property?

Seems like if a callback is provided, it shouldn't need any of the mechanism properties and therefore not should be required when a callback is provided.

Author: My initial intent was to define an API that would satisfy either a custom machine-to-machine callback or a built-in callback (e.g. the future built-in Azure OIDC logic, which will need extra configs to be provided by `authMechanismProperties`). On reflection, that seems overly prescriptive of the implementation rather than just the configuration API. I also intended `authMechanismProperties` to communicate additional configuration to the custom callback, but I realized that drivers validate the keys in `authMechanismProperties`, so passing arbitrary config values isn't possible.

As far as specifying the callback name in `authMechanismProperties`, I'm currently planning to drop that from the spec. I haven't been able to discover a use case for it and implementing that logic would be unconventional or nonsensical in many driver languages.

Author: Updated to not pass the mechanism properties to the callback and require that drivers enforce that `PROVIDER_NAME` and a custom callback cannot be specified at the same time. Also updated to support either specifying the callback via a `MongoClientConfig` property or a mechanism property, depending on which is supported or idiomatic for each driver.

machine-to-machine ("machine") and human-in-the-loop ("human"). Drivers MUST
support the machine authentication flow. Drivers MAY support the human
authentication flow.

Machine Authentication Flow
```````````````````````````
The machine authentication flow is intended to be used in cases where human
interaction is not necessary or practical, such as to authenticate database
access for a web service. Some OIDC documentation refers to the machine
authentication flow as "workload authentication".

Drivers MUST implement all behaviors described in the ``MONGODB-OIDC``
specification, unless the section or block specifically says that it only
applies to the `Human Authentication Flow`_.

Human Authentication Flow
`````````````````````````
The human authentication flow is intended to be used for applications that
involve direct human interaction, such as database tools or CLIs. Some OIDC
documentation refers to the human authentication flow as "workforce
authentication".

In addition to the **access token**, the human authentication flow introduces
some additional concepts:

- **Identity Provider (IdP)**: A service that manages user accounts and

>   authenticates users, such as Okta or OneLogin.

Author: Good point, I've added that clarification and moved the definitions to the top-level OIDC section.

Drivers MUST NOT send a ``PrincipalStepRequest`` when performing machine authentication
or when there is a cached Access Token.  Drivers MUST instead  use ``saslStart`` with a ``JwtStepRequest``.

Caching
```````
Drivers MUST cache the ``IdPInfo``, Access Token, and Refresh Token associated with each
``MongoClient``.  If any operation fails the driver MUST clear the Access Token.  The Refresh Token is handled differently,
see the ``Reauthentication`` section below.

Drivers MUST also use a token generation id that is incremented whenever a new Access Token is retrieved.
When a connection succeeds, the token generation id MUST be associated with or stored on the connection object.
This value is used in the ``Reauthentication`` section below.

Speculative Authentication
``````````````````````````
Drivers MUST implement speculative authentication for MONGODB-OIDC during the ``hello`` handshake.
If there is a cached Access Token the ``JwtStepRequest`` SASL command will be used as the speculation command.
If there is no cached Access Token, the ``PrincipalStepRequest`` will be used as the speculation command.

> The driver MUST NOT call the callback during speculative authentication.

Reviewer 3: I think we want two different callbacks, but I still think we shouldn't call it, because for Azure and perhaps other "custom" types there is an http request.

mechanism_properties
    PROVIDER_NAME
        Drivers MUST allow the user to specify the name of the OIDC service to

>         use to obtain credentials. MUST be one of ["aws"].

Reviewer: Is custom callback usage just the expected default here, if it can no longer be specified in the `PROVIDER_NAME` field?

Author: It's not the expected default. However, having to specify `PROVIDER_NAME="custom"` as well as specifying a custom callback seems redundant. Instead, I've updated the document to say that `PROVIDER_NAME` and a custom callback cannot be used together and drivers should return an error if both are specified.

Reviewer: Yeah, that clarifies things. Thank you!

Interfaces
``````````

Custom Callback
```````````````
Drivers MUST provide a way for the callback to be either
automatically canceled, or to cancel itself.  This can be as a timeout argument
to the callback, a cancellation context passed to the callback, or some other
language-appropriate mechanism.  The timeout deadline MUST be the same as the
connection establishment deadline (TODO: Does that make sense?).

Callbacks can be synchronous and/or asynchronous, depending on the driver
and/or language.  Asynchronous callbacks should be preferred when other
operations in the driver use asynchronous functions.

The driver MUST pass the following information to the callback:

- ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.
- ``mechanismProperties``: All key-value pairs from the
  AUTH_MECHANISM_PROPERTIES parameter.
- ``invalidatedToken``: Optional. The previous token that was invalidated after
  receiving a ``ReauthenticationRequired`` error.

> The callback MUST return an OIDC access token, formatted as a base64-encoded
> JWT.
> TODO: Is that correct?

Reviewer: Not sure whether this TODO refers to this specifically, but “base64-encoded JWT” might be confusing language; JWTs consist of period-separated base64url-encoded parts already by definition, so “base64-encoded JWT” sounds like the expectation is that there is an additional round of base64 encoding happening on the full JWT.

```suggestion
The callback MUST return an OIDC access token in JWT format.
```

Author: Agreed, simplified the wording based on your suggestion.

> The callback MUST return an OIDC access token, formatted as a base64-encoded
> JWT.
> TODO: Is that correct?

Reviewer: Not sure whether this TODO refers to this specifically, but “base64-encoded JWT” might be confusing language; JWTs consist of period-separated base64url-encoded parts already by definition, so “base64-encoded JWT” sounds like the expectation is that there is an additional round of base64 encoding happening on the full JWT.

```suggestion
The callback MUST return an OIDC access token in JWT format.
```

Reviewer: (sorry, not sure what was up with Github here)

The signature of the callback is up to the driver's discretion, but the driver
MUST ensure that additional optional parameters can be added to the callback
signature in the future. An example might look like:

.. code:: typescript

  interface CallbackParameters {
      // All key-value pairs from the AUTH_MECHANISM_PROPERTIES parameter.

>       mechanismProperties Object;

Reviewer: ```suggestion
      mechanismProperties: Object;
```

>       mechanismProperties Object;

Reviewer: ```suggestion
      mechanismProperties: Object;
```

mechanism_properties
    PROVIDER_NAME
        The name of a built-in OIDC provider integration to use to obtain
        credentials. The value MUST be one of ["aws"]. If both ``PROVIDER_NAME``
        and an `OIDC Callback`_ are provided for the same ``MongoClient``
        (either via the ``CALLBACK`` mechanism property or a ``MongoClient``
        configuration), the driver MUST raise an error.

    CALLBACK
        An `OIDC Callback` that returns credentials for OIDC providers that do
        not have a built-in integraiton. Drivers MAY allow the user to specify
        an `OIDC Callback`_ using a ``MongoClient`` configuration instead of a
        mechanism property, depending on what is conventional for the driver.
        Drivers MUST NOT support both the ``CALLBACK`` mechanism property and
        the ``MongoClient`` configuration.

    CALLBACK_TYPE
        The type of `OIDC Callback`. The value MUST be one of ["machine",
        "human"]. If an `OIDC Callback`_ is configured and ``CALLBACK_TYPE`` is
        not specified, the driver MUST raise an error. This property is only

>         required for drivers that support the `Human Authentication Flow`_.

Author: Yeah, that makes sense. It would be weird if they were specified in different places. I also added a note that drivers can use a `MongoClient` config to specify the callback type if they also use a `MongoClient` config to specify the callback.

>         required for drivers that support the `Human Authentication Flow`_.

Author: Based on Jeff's feedback, my plan is to actually remove `CALLBACK_TYPE` and replace it with either a separate callback argument specifically for the human callback or a boolean value to enable the human auth flow. In either case, I will still require that it not be specified in the connection string.

drivers MUST be able to pass the following information to the callback:

- ``idpInfo``: Information used to authenticate with the IdP.

  - ``issuer``: A URL which describes the Authentication Server. This identifier
    should be the iss of provided access tokens, and be viable for RFC8414
    metadata discovery and RFC9207 identification.
  - ``clientId``: A unique client ID for this OIDC client.
  - ``requestScopes``: A list of additional scopes to request from IdP.

- ``refreshToken``: The refresh token, if applicable, to be used by the callback
  to request a new token from the issuer.

In addition to the information described in the `OIDC Callback` section, the
callback MUST be able to return the following information:

- ``refreshToken``: An optional refresh token that can be used to fetch new
  access tokens.

An example callback API that supports the human callback might look like:

.. code:: typescript

  interface IdpInfo {
    issuer: string;
    clientId: string;
    requestScopes: Optional<Array<string>>;
  }

  interface OIDCCallbackParams {
      callbackTimeoutMS: int;
      version: int;
      idpInfo: Optional<IdpInfo>;

>       refreshToken: Optional<string>;

Reviewer: Do we enforce this being a string? From the driver's perspective, it can be any opaque value (whatever that looks like in a given language), right?

Author: No, it can be any value. I changed the type to `any`.

    MUST be "MONGODB-OIDC"

mechanism_properties
  PROVIDER_NAME
    Drivers MUST allow the user to specify the name of a built-in OIDC provider
    integration to use to obtain credentials. If provided, the value MUST be one

>     of ["aws"]. If both ``PROVIDER_NAME`` and an `OIDC Callback`_ or `Human

Reviewer 4: Should we call `Human Callback` `OIDC Human Callback` to be consistent with the mechanism property name?

mechanism_properties
    PROVIDER_NAME
        The name of a built-in OIDC provider integration to use to obtain
        credentials. The value MUST be one of ["aws"]. If both ``PROVIDER_NAME``
        and an `OIDC Callback`_ are provided for the same ``MongoClient``
        (either via the ``CALLBACK`` mechanism property or a ``MongoClient``
        configuration), the driver MUST raise an error.

    CALLBACK
        An `OIDC Callback`_ that returns credentials for OIDC providers that do
        not have a built-in integraiton. Drivers MAY allow the user to specify
        an `OIDC Callback`_ using a ``MongoClient`` configuration instead of a
        mechanism property, depending on what is conventional for the driver.
        Drivers MUST NOT support both the ``CALLBACK`` mechanism property and
        the ``MongoClient`` configuration.

>     CALLBACK_TYPE

Reviewer 4: This property seems odd. Consider a MongoClient configuration for the OIDC callback: what should its signature be? For statically typed languages, you would define a type for each callback.  In Java, you might have:

```
interface HumanOidcCallback {
   HumanCallbackReturnType callback(HumanCallbackParams);
}

interface MachineOidcCallback {
   MachineCallbackReturnType callback(MachineCallbackParams);
}

...
   void oidcCallbackType(String type) ...
   void registerOidcCallback(...  What type should this be? In Java the only choice would be Object
```

I think it would be better, at least in Java, to instead overload the second method:

```
   void registerOidcCallback(HumanOidcCallback callback) ...
   void registerOidcCallback(MachineOidcCallback callback) ...
```

and it would be a runtime error to register both, and then the callback type is not necessary.

Also, see my comment below on callback type not being required in all cases, as it's related.

Author: In the first example, both callbacks would use the same interface, so there would only be one `registerOidcCallback` method that accepted the `OidcCallback` type. The params and return values for the human auth flow are a superset of the params and return types for the machine auth flow. To support the human auth flow, `OidcCallback` would include all params and return values for both the machine and human auth flows, and `oidcCallbackType` would specify which flow to use.

However, I see your point about the more idiomatic pattern in Java being to use method overriding to set the appropriate OIDC callback. Seems like the pattern for passing callbacks can be very different in different drivers and languages, so I think the current language in the callback API section is probably too prescriptive. I will update that section to be more permissive so drivers can implement a callback API that is idiomatic for their driver and language.

        The type of `OIDC Callback`_. The value MUST be one of ["machine",
        "human"]. Drivers MUST NOT allow the user to specify ``CALLBACK_TYPE``
        in the connection string. Drivers MAY allow the user to specify the
        callback type using a ``MongoClient`` configuration instead of a
        mechanism property to be consistent with how the `OIDC Callback`_ is
        configured. If an `OIDC Callback`_ is configured and ``CALLBACK_TYPE``
        is not specified, the driver MUST raise an error. This property is only

>         required for drivers that support the `Human Authentication Flow`_.

Reviewer 4: I think we have to say that the property is required in all cases.  Otherwise introduction of the human flow in a subsequent release will be a breaking change, because an existing app that has registered a machine callback will error on upgrade.

Author: Good point, that would create a breaking change for drivers that add human auth flow support later. Based on that and your other comments, I think we should not use `CALLBACK_TYPE`. Instead, we should replace it with either a property like `USE_HUMAN_CALLBACK:true` or a property like `HUMAN_CALLBACK` that expects a callback function and is mutually exclusive with `CALLBACK`.

Reviewer 4: Resolved by adding OIDC_HUMAN_CALLBACK

mechanism_properties
    PROVIDER_NAME
        MUST NOT be specified.

    REQUEST_TOKEN_CALLBACK
        Drivers MUST allow the user to specify a callback of the form "onRequest" (defined below), if the driver supports
        providing objects as mechanism property values.  Otherwise the driver MUST allow it as a MongoClientOption.

    ALLOWED_HOSTS
        The list of allowed hostnames or ip-addresses (ignoring ports) for
        MongoDB connections. The hostnames may include a leading "\*." wildcard, which allows for matching
        (potentially  nested) subdomains. ALLOWED_HOSTS is a
        security feature and MUST default to

>         ``["*.mongodb.net", '*.mongodb-qa.net", '*.mongodb-dev.net", "*.mongodbgov.net", "localhost", "127.0.0.1", "::1"]``.

Reviewer 5: ```suggestion
        ``["*.mongodb.net", "*.mongodb-qa.net", "*.mongodb-dev.net", "*.mongodbgov.net", "localhost", "127.0.0.1", "::1"]``.
```

If the callback does not return an object in the correct form of ``IdpResponse``, the driver MUST raise an error either using the type
system or by raising an error when non-optional properties are missing .
The driver MUST NOT attempt to validate the token(s) directly.
It is expected that if the server changes the expected fields, the SASL exchange will be updated with a version parameter.
Drivers do not need to attempt to provide old-driver-new-server compatibility.

If no callback is given, the driver MUST must raise a validation error.

Reauthentication
````````````````

When reauthentication is requested by the server (as a 391 error code) and MONGODB-OIDC is in use, the driver MUST perform a reauthentication.
The driver MUST account for the case of multiple connections hitting a reauthentication error at different times.
The driver MUST also account for a reauthenication that results from an IdP configuration change on the server.
To accomplish these goal, the following algorithm is used to handle a reauthenication error:

- First, see if the Access Token on the MongoClient is different than the Access Token on the connection.

>    - If they are different, optimisitically try to authenticate.  On error, continue.

Reviewer 5: ```suggestion
   - If they are different, optimisitically try to authenticate.  On error, continue.
```

Can we clear up exactly what this means? Does this mean to use the accessToken from the client and set it on the connection? Does it mean to use the refresh token? Do we need to cache the conversationId from the original authentication flow?

Author: Rewrote this whole section to be clearer.

- Check if the the *Client Cache* has an access token.
    - If it does, cache the returned access token in *Connection Cache* and
      optimisitically try to authenticate using the access token. If the server

>       returns ``AuthenticationFailed`` (error code 18), slep 100ms then

Author: Fixed!
